src/routes/mod_creation.rs

```diff
@@ -0,0 +1,401 @@
+pub async fn mod_create(
+    mut payload: Multipart,
+    client: Data<Client>,
+    upload_url: Data<UploadUrlData>,
+) -> Result<HttpResponse, CreateError> {
+    //TODO Switch to transactions for safer database and file upload calls (once it is implemented in the APIs)
+    let cdn_url = dotenv::var("CDN_URL")?;
+
+    let db = client.database("modrinth");
+
+    let mods = db.collection("mods");
+    let versions = db.collection("versions");
+
+    let mut mod_id = ModId(random_base62(8));
+
+    //Check if ID is unique
+    loop {
+        let filter = doc! { "_id": mod_id.0 };
+
+        if mods.find(filter, None).await?.next().await.is_some() {
+            mod_id = ModId(random_base62(8));
+        } else {
+            break;
+        }
+    }
+
+    let mut created_versions: Vec<Version> = vec![];
+
+    let mut mod_create_data: Option<ModCreateData> = None;
+    let mut icon_url = "".to_string();
+
+    let mut current_file_index = 0;
+    while let Some(item) = payload.next().await {
+        let mut field: Field = item.map_err(CreateError::MultipartError)?;
+        let content_disposition =
+            field
+                .content_disposition()
+                .ok_or(CreateError::MissingValueError(
+                    "Missing content disposition!".to_string(),
+                ))?;
+        let name = content_disposition
+            .get_name()
+            .ok_or(CreateError::MissingValueError(
+                "Missing content name!".to_string(),
+            ))?;
+
+        while let Some(chunk) = field.next().await {
+            let data = &chunk.map_err(CreateError::MultipartError)?;
+
+            if name == "data" {
+                mod_create_data = Some(serde_json::from_slice(&data)?);
+            } else {
+                let file_name =
+                    content_disposition
+                        .get_filename()
+                        .ok_or(CreateError::MissingValueError(
+                            "Missing content file name!".to_string(),
+                        ))?;
+                let file_extension = String::from_utf8(
+                    content_disposition
+                        .get_filename_ext()
+                        .ok_or(CreateError::MissingValueError(
+                            "Missing file extension!".to_string(),
+                        ))?
+                        .clone()
+                        .value,
+                )
+                .map_err(CreateError::InvalidUtf8Input)?;
+
+                if let Some(create_data) = &mod_create_data {
+                    if name == "icon" {
+                        if let Some(ext) = get_image_content_type(file_extension) {
+                            let upload_data = upload_file(
+                                upload_url.get_ref().clone(),
+                                ext,
+                                format!("mods/icons/{}/{}", mod_id.0, file_name),
+                                data.to_vec(),
```

Comment:
Uploading a file should borrow a slice rather than requiring an extra allocation.

Reply:
Will fix this.

Reply:
Okay, I figured out why I didn't do this earlier, reqwest doesn't take in a slice for it's body for some reason. (they do but only one with a static lifetime)